File: src/agent/providers/transport.ts
import { ProviderError, ProviderStreamEvent } from "../types";

/**
 * The pieces every provider adapter needs from the network, in one place so the
 * four of them can't drift on the parts that have already cost debugging time:
 * how cancellation is detected, and how an SSE byte stream becomes events.
 */

/**
 * The subset of `fetch` the adapters need, kept structural on purpose: it has
 * to accept the platform `fetch`, `expo/fetch` (whose `Response` is its own
 * class, not the DOM one), and a plain stub in tests — without the agent core
 * importing anything platform-specific to describe them.
 */
export type FetchLikeInit = {
  method?: string;
  headers?: Record<string, string>;
  body?: string;
  signal?: AbortSignal;
};

export type FetchLikeResponse = {
  ok: boolean;
  status: number;
  /** Null whenever the implementation can't stream, which the adapters report rather than hiding. */
  body: ReadableStream<Uint8Array> | null;
  json: () => Promise<unknown>;
};

export type FetchLike = (url: string, init?: FetchLikeInit) => Promise<FetchLikeResponse>;

function resolveFetch(injected?: FetchLike): FetchLike | undefined {
  return injected ?? (typeof globalThis.fetch === "function" ? (globalThis.fetch as unknown as FetchLike) : undefined);
}

export function toTransportError(err: unknown, signal?: AbortSignal): ProviderError {
  // An aborted signal is the reliable evidence, not the error's shape. Every
  // transport words cancellation differently — the DOM throws an `AbortError`,
  // while `expo/fetch` throws `Expo.FetchRequestCanceledException`, whose name
  // matches nothing — and matching on names leaked a raw native message to the
  // user whenever a new transport appeared.
  //
  // An adapter still can't know *why* the signal fired: a user tapping cancel
  // and a deadline elapsing look identical here. Report the neutral fact and let
  // the caller, which owns the abort reason, decide how to describe it.
  if (signal?.aborted || (err instanceof Error && err.name === "AbortError")) {
    return { kind: "cancelled", message: "The request was cancelled." };
  }
  return { kind: "network", message: err instanceof Error ? err.message : "Network request failed." };
}

/** HTTP status → error kind. Shared so one provider can't quietly disagree about what 429 means. */
export function httpErrorKind(status: number): ProviderError["kind"] {
  if (status === 401 || status === 403) return "auth";
  if (status === 402) return "insufficient_credits";
  if (status === 429) return "rate_limit";
  if (status === 404) return "not_found";
  return "unknown";
}

/**
 * Whether a 400 is the provider saying this model can't use tools *here*.
 *
 * Status alone can't tell: every "you asked for something this model won't do"
 * arrives as a generic 400 with the reason only in prose, so this reads the
 * message. That is a heuristic and deliberately a narrow one — it requires both
 * a mention of tools and an explicit refusal, and the provider's own message is
 * always kept alongside the kind so a miss degrades to showing the real text
 * rather than to a wrong diagnosis.
 *
 * The case that motivated it: OpenAI's reasoning models reject tools on
 * /v1/chat/completions entirely ("Function tools with reasoning_effort are not
 * supported for <model> in /v1/chat/completions"), which surfaced to the user as
 * a raw API sentence with no hint that picking another model would fix it.
 */
export function isToolSupportError(status: number, message: string): boolean {
  if (status !== 400) return false;
  const m = message.toLowerCase();
  const mentionsTools = m.includes("tool") || m.includes("function");
  const refuses = m.includes("not supported") || m.includes("unsupported") || m.includes("does not support");
  return mentionsTools && refuses;
}

/**
 * Whether a 400 is the provider saying it has never heard of this model.
 *
 * Most endpoints answer an unknown model with a 404, which needs no help.
 * OpenAI's `/v1/responses` answers with a **400** — "The requested model 'x'
 * does not exist." — so without this a typo'd model id reads as a generic
 * failure and the user is shown raw API prose instead of being sent to the
 * model field in Settings.
 */
export function isModelNotFoundError(status: number, message: string): boolean {
  if (status !== 400) return false;
  const m = message.toLowerCase();
  if (!m.includes("model")) return false;
  return m.includes("does not exist") || m.includes("not found") || m.includes("unknown model");
}

/**
 * The kind for a failed HTTP response, given both the status and what the body
 * said. Prefer this to `httpErrorKind` at an adapter's error boundary: status is
 * enough for the well-known codes, but not for the 400s that only differ in prose.
 *
 * Tool support is checked first: "function tools ... are not supported for
 * <model>" names a model too, and the useful advice there is to change model
 * because of what it can't do, not because it doesn't exist.
 */
export function classifyHttpError(status: number, message: string): ProviderError["kind"] {
  if (isToolSupportError(status, message)) return "no_tool_support";
  if (isModelNotFoundError(status, message)) return "not_found";
  return httpErrorKind(status);
}

/**
 * Reads an error body without assuming a shape. OpenAI-compatible endpoints and
 * Anthropic both nest under `error.message`; Google returns `error.message` too
 * but sometimes as an array of details. Anything unrecognised falls back to the
 * caller's generic message rather than showing the user raw JSON.
 */
async function readErrorMessage(response: FetchLikeResponse, fallback: string): Promise<string> {
  try {
    const json = (await response.json()) as { error?: { message?: unknown } | string };
    if (typeof json?.error === "string" && json.error.length > 0) return json.error;
    const message = (json?.error as { message?: unknown })?.message;
    if (typeof message === "string" && message.length > 0) return message;
  } catch {
    // Body wasn't JSON — keep the generic message.
  }
  return fallback;
}

/**
 * A usage event, or null when there is nothing worth reporting.
 *
 * Shared because the adapters had drifted: two suppressed an all-zero reading
 * and one emitted it, so the same silent stream produced different event
 * sequences depending on the provider. Nothing consumes usage yet — per-turn
 * cost is deferred (PRD §13) — which is the reason to settle the shape now,
 * while no caller depends on it.
 */
export function usageEvent(input: unknown, output: unknown): ProviderStreamEvent | null {
  const inputTokens = Number(input ?? 0);
  const outputTokens = Number(output ?? 0);
  if (!Number.isFinite(inputTokens) || !Number.isFinite(outputTokens)) return null;
  if (inputTokens + outputTokens <= 0) return null;
  return { type: "usage", inputTokens, outputTokens };
}

/**
 * The request every adapter makes, and the three failures they all share:
 * no fetch was injected, the response was not ok, and the transport returned a
 * body that can't be streamed.
 *
 * Each adapter had its own copy of this — same logic, same wording, three
 * chances to drift. It resolves to the stream to parse, or the single error
 * event to yield instead, so the caller stays a straight line.
 */
export async function openProviderStream(opts: {
  url: string;
  init: FetchLikeInit;
  /** Names the provider in error text, so a failure blames whoever actually failed. */
  displayName: string;
  fetch?: FetchLike;
}): Promise<{ stream: ReadableStream<Uint8Array> } | { error: ProviderError }> {
  const doFetch = resolveFetch(opts.fetch);
  if (!doFetch) {
    return {
      error: { kind: "network", message: "No fetch implementation available. Pass one via the provider options." },
    };
  }

  let response: FetchLikeResponse;
  try {
    response = await doFetch(opts.url, opts.init);
  } catch (err) {
    return { error: toTransportError(err, opts.init.signal) };
  }

  if (!response.ok) {
    const message = await readErrorMessage(response, `${opts.displayName} request failed (HTTP ${response.status}).`);
    return { error: { kind: classifyHttpError(response.status, message), message, status: response.status } };
  }

  // A successful response with no readable body means the fetch in use can't
  // stream. Reporting that as an HTTP error would blame the provider for a local
  // wiring problem ("request failed (HTTP 200)").
  if (!response.body) {
    return {
      error: {
        kind: "network",
        message:
          "The fetch implementation in use does not expose a streaming response body, so the model's reply cannot be read. Pass expo/fetch to the provider.",
      },
    };
  }

  return { stream: response.body };
}

/** Matches an SSE event separator: a blank line, in either LF or CRLF form. */
const EVENT_SEPARATOR = /\r?\n\r?\n/;

function splitFirstEvent(buffer: string): { event: string; rest: string } | null {
  const match = EVENT_SEPARATOR.exec(buffer);
  if (!match) return null;
  return { event: buffer.slice(0, match.index), rest: buffer.slice(match.index + match[0].length) };
}

/**
 * Yields each raw SSE event block from a byte stream.
 *
 * Two behaviours here were bugs once and must not regress: events can be split
 * across chunk boundaries, so the buffer is only consumed at separators; and a
 * stream that ends *without* a trailing blank line still holds one real event,
 * which is dropped if the buffer isn't flushed — losing the model's last text
 * delta, or a whole tool call.
 *
 * Read failures are deliberately allowed to propagate: aborting mid-stream
 * rejects the reader, and only the caller knows whether that was a cancellation.
 */
export async function* readSseEvents(stream: ReadableStream<Uint8Array>): AsyncIterable<string> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      let split = splitFirstEvent(buffer);
      while (split) {
        buffer = split.rest;
        yield split.event;
        split = splitFirstEvent(buffer);
      }
    }

    buffer += decoder.decode();
    if (buffer.trim().length > 0) yield buffer;
  } finally {
    reader.releaseLock();
  }
}

/**
 * The `data:` payload of one SSE event, or null for anything without one —
 * keep-alive comments (`: ping`), bare `event:` lines, and the `[DONE]`
 * terminator all arrive as events with nothing to parse.
 */
export function sseData(event: string): string | null {
  const lines = event.split(/\r?\n/);
  const parts: string[] = [];
  for (const line of lines) {
    if (line.startsWith("data:")) parts.push(line.slice(5).trim());
  }
  if (parts.length === 0) return null;
  const data = parts.join("\n");
  return data === "[DONE]" || data.length === 0 ? null : data;
}

/** The `event:` name of an SSE event, which Anthropic relies on and OpenAI never sends. */
export function sseEventName(event: string): string | null {
  for (const line of event.split(/\r?\n/)) {
    if (line.startsWith("event:")) return line.slice(6).trim();
  }
  return null;
}

File: lib/features/ai/data/agent/transport.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'types.dart';

/// The response shape adapters work against, structural so tests can fake it.
abstract interface class HttpResponseLike {
  bool get ok;
  int get status;

  /// Null when the implementation cannot stream — adapters then report it
  /// rather than pretending.
  Stream<List<int>>? get body;

  /// Reads the whole body as text (error payloads, non-streaming responses).
  Future<String> text();
}

class _StreamedHttpResponse implements HttpResponseLike {
  _StreamedHttpResponse(this._response);

  final http.StreamedResponse _response;

  @override
  bool get ok => _response.statusCode >= 200 && _response.statusCode < 300;

  @override
  int get status => _response.statusCode;

  @override
  Stream<List<int>>? get body => _response.stream;

  @override
  Future<String> text() => _response.stream.bytesToString();
}

/// The tiny slice of the network transport the provider adapters need,
/// injectable for tests. The app wires the real implementation in the
/// composition root.
abstract interface class HttpTransport {
  Future<HttpResponseLike> post(
    Uri url, {
    required Map<String, String> headers,
    required String body,
    CancellationToken? token,
  });
}

class HttpTransportImpl implements HttpTransport {
  HttpTransportImpl([http.Client? client]) : _client = client ?? http.Client();

  final http.Client _client;

  @override
  Future<HttpResponseLike> post(
    Uri url, {
    required Map<String, String> headers,
    required String body,
    CancellationToken? token,
  }) async {
    token?.throwIfCancelled();
    final request = http.Request('POST', url)
      ..headers.addAll(headers)
      ..body = body;
    final response = await _client.send(request);
    return _StreamedHttpResponse(response);
  }
}

/// Transports the status into the kind, shared so one provider can't quietly
/// disagree about what 429 means.
ProviderErrorKind httpErrorKind(int status) {
  if (status == 401 || status == 403) return ProviderErrorKind.auth;
  if (status == 402) return ProviderErrorKind.insufficientCredits;
  if (status == 429) return ProviderErrorKind.rateLimit;
  if (status == 404) return ProviderErrorKind.notFound;
  return ProviderErrorKind.unknown;
}

/// Whether a 400 is the provider saying this model can't use tools *here*.
///
/// Status alone can't tell — the "you asked for something this model won't do"
/// failures arrive as a generic 400 with the reason only in prose. This reads
/// the message and requires both a tool mention and an explicit refusal so a
/// miss degrades to showing the real text instead of a wrong diagnosis.
bool isToolSupportError(int status, String message) {
  if (status != 400) return false;
  final m = message.toLowerCase();
  final mentionsTools = m.contains('tool') || m.contains('function');
  final refuses = m.contains('not supported') ||
      m.contains('unsupported') ||
      m.contains('does not support');
  return mentionsTools && refuses;
}

/// Whether a 400 is the provider saying it has never heard of this model
/// (OpenAI's `/v1/responses` answers unknown models with a 400, not a 404).
bool isModelNotFoundError(int status, String message) {
  if (status != 400) return false;
  final m = message.toLowerCase();
  if (!m.contains('model')) return false;
  return m.contains('does not exist') ||
      m.contains('not found') ||
      m.contains('unknown model');
}

/// The kind for a failed HTTP response, given both the status and what the
/// body said. Prefer this over [httpErrorKind] at an adapter's error
/// boundary: status is enough for the well-known codes, but not for the 400s
/// that only differ in prose.
ProviderErrorKind classifyHttpError(int status, String message) {
  if (isToolSupportError(status, message)) {
    return ProviderErrorKind.noToolSupport;
  }
  if (isModelNotFoundError(status, message)) {
    return ProviderErrorKind.notFound;
  }
  return httpErrorKind(status);
}

/// Reads a provider's error message without assuming a shape. OpenAI-shaped
/// endpoints and Anthropic nest under `error.message`; Google returns
/// `error.message` too. Anything unrecognised falls back to the caller's
/// generic message.
String errorMessageFromBody(String body, String fallback) {
  try {
    final decoded = jsonDecode(body);
    if (decoded is Map<String, dynamic>) {
      final message = decoded['error'];
      if (message is Map<String, dynamic> && message['message'] is String) {
        return message['message'] as String;
      }
      if (message is String && message.isNotEmpty) return message;
    }
  } catch (_) {
    // fall through
  }
  return fallback;
}

/// Classifies a transport failure. An aborted token is the reliable evidence,
/// not the error's shape — every transport words cancellation differently.
ProviderError toTransportError(Object? error, CancellationToken? token) {
  if (token?.isCancelled == true || error is CancelledException) {
    return const ProviderError(
        kind: ProviderErrorKind.cancelled,
        message: 'The request was cancelled.');
  }
  return ProviderError(
    kind: ProviderErrorKind.network,
    message: error is Exception ? error.toString() : 'Network request failed.',
  );
}

UsageEvent? usageEvent(int? inputTokens, int? outputTokens) {
  if (inputTokens == null || outputTokens == null) return null;
  return UsageEvent(inputTokens: inputTokens, outputTokens: outputTokens);
}

class SseEvent {
  const SseEvent({required this.event, required this.data});

  /// Empty when the stream only carries `data:` frames.
  final String event;
  final String data;
}

/// Splits an SSE byte stream into events, decoding one complete line at a
/// time so a multi-byte UTF-8 character can never be split across chunk
/// boundaries (`\n` is not a UTF-8 continuation byte).
///
/// [token] is checked between chunks so cancellation stops reading promptly.
Stream<SseEvent> readSseEvents(Stream<List<int>> bytes,
    {CancellationToken? token}) async* {
  final decoder = const Utf8Decoder(allowMalformed: true);
  var pending = <int>[];
  final readyEvents = <SseEvent>[];
  var dataLines = <String>[];
  var eventType = '';

  void processLine(List<int> rawLine) {
    var line = decoder.convert(rawLine);
    if (line.endsWith('\r')) {
      line = line.substring(0, line.length - 1);
    }
    if (line.isEmpty) {
      if (dataLines.isNotEmpty) {
        readyEvents.add(SseEvent(event: eventType, data: dataLines.join('\n')));
        dataLines = <String>[];
        eventType = '';
      }
      return;
    }
    if (line.startsWith(':')) return; // comment
    if (line.startsWith('data:')) {
      var value = line.substring(5);
      if (value.startsWith(' ')) value = value.substring(1);
      dataLines.add(value);
      return;
    }
    if (line.startsWith('event:')) {
      eventType = line.substring(6).trim();
    }
  }

  await for (final chunk in bytes) {
    token?.throwIfCancelled();
    pending.addAll(chunk);
    while (true) {
      final nl = pending.indexOf(0x0A);
      if (nl == -1) break;
      processLine(pending.sublist(0, nl));
      pending = pending.sublist(nl + 1);
    }
    while (readyEvents.isNotEmpty) {
      final next = readyEvents.removeAt(0);
      yield next;
    }
  }

  // Flush anything after the final newline, then a data-only event with no
  // terminating blank line.
  if (pending.isNotEmpty) processLine(pending);
  if (dataLines.isNotEmpty) {
    yield SseEvent(event: eventType, data: dataLines.join('\n'));
  }
  while (readyEvents.isNotEmpty) {
    final next = readyEvents.removeAt(0);
    yield next;
  }
}

/// Shortcut for classifying a failed response into the error the loop reports.
ProviderError classifiedError(int status, String message, String fallback) =>
    ProviderError(
      kind: classifyHttpError(status, message),
      message: message.isNotEmpty ? message : fallback,
      status: status,
    );
